e2e_test/js-tests/test_viem_smoketest.mjs

```diff
@@ -0,0 +1,170 @@
+	account,
+	chain: chain,
+	transport: http(),
+});
+
+// check checks that the receipt has status success and that the transaction
+// type matches the expected type, since viem sometimes mangles the type when
+// building txs.
+async function check(txHash, type) {
+	const receipt = await publicClient.waitForTransactionReceipt({ hash: txHash });
+	const transaction = await publicClient.getTransaction({ hash: txHash });
+	assert.equal(transaction.type, type, "transaction type does not match");
+	assert.equal(receipt.status, "success", "receipt status 'failure'");
+}
+
+// sendTypedTransaction sends a transaction with the given type and an optional
+// feeCurrency.
+async function sendTypedTransaction(type, feeCurrency){
+		 return await walletClient.sendTransaction({
+			to: "0x00000000000000000000000000000000DeaDBeef",
+			value: 1,
+			type: type,
+			feeCurrency: feeCurrency,
+		});
+}
+
+// sendTypedCreateTransaction sends a create transaction with the given type
+// and an optional feeCurrency.
+async function sendTypedCreateTransaction(type, feeCurrency){
+		 return await walletClient.deployContract({
+			type: type,
+			feeCurrency: feeCurrency,
+			bytecode: testContractJSON.bytecode.object,
+			abi: testContractJSON.abi,
+			 // The constructor args for the test contract at ../debug-fee-currency/DebugFeeCurrency.sol
+			args:[1n, true, true, true],
+		});
+}
+
+// verifyTypedTransactions is a helper function that submits a send and create
+// transaction of the given type with optional feeCurrency and checks the
+// results.
+async function verifyTypedTransactions(type, feeCurrency){
+		const send = await sendTypedTransaction(type, feeCurrency);
+		await check(send, type);
+		const create = await sendTypedCreateTransaction(type, feeCurrency);
+		await check(create, type);
+}
+
+describe("viem smoke test", () => {
+	it("send legacy tx", async () => {
+		const type = "legacy";
+		await verifyTypedTransactions(type);
+	}).timeout(30_000);
+
+	it("send eip2930 tx", async () => {
+		const type = "eip2930";
+		await verifyTypedTransactions(type);
+	}).timeout(10_000);
+
+	it("send eip1559 tx", async () => {
+		const type = "eip1559";
+		await verifyTypedTransactions(type);
+	}).timeout(10_000);
+
+	it("send cip64 tx", async () => {
+		const type = "cip64";
+		await verifyTypedTransactions(type, process.env.FEE_CURRENCY);
+	}).timeout(10_000);
+
+	// This test is failing because the produced transaction is of type cip64.
+	// I guess this is a problem with the viem internals.
+	it.skip("cip42 not supported", async () => {
+		const type = "cip42";
+		await verifyTypedTransactions(type);
+	}).timeout(10_000);
+
+	it.only("legacy tx with fee currency not supported", async () => {
+		const type = "legacy";
+		try {
+			const hash = await sendTypedTransaction(type,process.env.FEE_CURRENCY);
+
+			// When using the devnet an exception is thrown from
+			// sendTypedTransaction, on alfajores the transaction is submitted so we
```

Comment:
Similar to the comment with the skipped test above - shouldn't this be caught early by viem and refuse to send unsupported type/currency combinations?

Reply:
It should be yes

Reply:
the serializer only knows what it knows. If the transaction fit as the celo specific one then it delegates to the serializer for op chains which in turn delegates to the default serializer. We wouldnt want to block unknowns configurations as they could known by the other serializer. 

Our focus on viem was getting the cip64 type to work. not blocking unsupported transactions types. 


If you put "legacy" as the type if anything it will take that to mean eth type 0 transaction not the celo legacy. 


I think the use case you are building here is pretty unique, i doubt anyone else tries this. 


If you wish to test sending txns to the blockchain one option would be to create your own serializers 

```ts
function serializeBaddies(transaction) {
 if  (type == 'celo-legacy')  {
 const rlpEncode = toRlp([
    stringNumberToHex(transaction.nonce),
    transaction.gasPrice,
    transaction.gas,
    transaction.feeCurrency,
    transaction.gatewayFeeRecipient,
    transaction.gatewayFee,
    transaction.to,
    transaction.value,
    transaction.data,
    stringNumberToHex(transaction.chainId),
    '0x',
    '0x',
  ])
 } else if type === 'cip42' {
  ...
 
 }
 else
    return  serializeTransactionImporFromCelo(tx)
}
```

Reply:
then add that serializer to the chain def 

`const celoWithOldTxns = defineChain({...celo, serializer: serializeBaddies})`

Reply:
Thanks for the clarification @aaronmgdr!

Reply:
So rather than trying to bend viem into sending transactions it shouldn't I've added a golang test that sends unsupported transactions and checks that they are not accepted by the node.